File: assign3/kotlin/lib/bin/main/org/example/Scheduler.kt
package org.example

class MeetingDetails {
  val with = this
  val that = this
  val on = this
  val and = this
  val participants = this
  
  var meetingName = ""
  var meetingStartTime = -1
  var meetingEndTime = -1
  var meetingDay = -1
  var meetingMonth = ""
  var meetingYear = -1
  var meetingParticipants = ""

  infix fun name(meetingName: String) {
    this.meetingName = meetingName
  }

  infix fun starts(time: Int) {
    meetingStartTime = time
  }

  infix fun ends(time: Int) {
    meetingEndTime = time
  }

  infix fun date(day: Int): MeetingDetails {
    meetingDay = day
    return this
  }

  infix fun October(year: Int) {
    meetingMonth = "October"
    meetingYear = year
  }

  infix fun are(participants: String){
    meetingParticipants = participants
  }

  private fun formatTime(time: Int) = if (time >= 0) "%02d:%02d".format(time / 100, time % 100) else ""

  private fun pluralize(unit: String, quantity: Int) = "$quantity $unit${if (quantity != 1) "s" else ""}"

  private fun calculateDuration(): String {
    if (meetingStartTime < 0 || meetingEndTime < 0) return ""
    
    val minutes = ((meetingEndTime / 100) * 60 + meetingEndTime % 100) - ((meetingStartTime / 100) * 60 + meetingStartTime % 100)
    return "Duration: ${pluralize("hour", minutes / 60)} and ${pluralize("minute", minutes % 60)}"
  }

  fun buildMeetingString(): String = buildString {
    append("Meeting: $meetingName")
    formatTime(meetingStartTime).takeIf { it.isNotEmpty() }?.let { append(", Start Time: $it") }
    formatTime(meetingEndTime).takeIf { it.isNotEmpty() }?.let { append(", End Time: $it") }
    if (meetingStartTime >= 0 && meetingEndTime >= 0) append(", ${calculateDuration()}")
    if (meetingDay != -1 && meetingMonth.isNotEmpty() && meetingYear != -1) append(", Date: $meetingMonth $meetingDay $meetingYear")
    if (meetingParticipants.isNotEmpty()) append(", Participants: $meetingParticipants")
  }
    
}

object schedule {
  infix fun meeting(block: MeetingDetails.() -> Unit) = MeetingDetails().apply(block).buildMeetingString()
}

File: assign3/ruby/scheduler.rb
class MeetingDetails
  attr_reader :name, :starts, :ends, :on, :participating

  def name(name)
    @name = name
  end

  def set_time(time)
    [time.div(100), time % 100]
  end

  def starts(time)
    @starts = set_time(time)
  end

  def ends(time)
    @ends = set_time(time)
  end

  def duration
    return unless @starts && @ends
    
    hours, minutes = ((@ends.first * 60 + @ends.last) - (@starts.first * 60 + @starts.last)).divmod(60)
  
    "Duration: #{hours} hour#{'s' if hours != 1} and #{minutes} minute#{'s' if minutes != 1}"
  end

  def on(on)
    @on = on
  end

    Integer.class_eval do
      def /(n)
        "#{self}/#{n}"
      end
    end
    
    String.class_eval do
      def /(n)
        "#{self}/#{n}"
      end
    end

  def participating(*participants)
    @participating = participants
  end

  def to_s
    meeting = "Meeting: #{@name}"
    meeting += ", Start Time: #{@starts.first}:#{@starts.last}" if @starts
    meeting += ", End Time: #{@ends.first}:#{@ends.last}" if @ends
    meeting += ", #{duration}" if @starts && @ends
    meeting += ", Date: #{@on}" if @on
    meeting += ", #{@participating.size} Participants, Participants: #{@participating.join(', ')}" if @participating
    meeting
  end

end

def schedule(meeting)
  meeting
end

def meeting(&block)
  context = MeetingDetails.new
  
  context.instance_eval(&block) if block_given?
  
  context.to_s
end
